request_helpers.go

```diff
@@ -0,0 +1,187 @@
+package linodego
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/url"
+	"strconv"
+)
+
+// paginatedResponse represents a single response from a paginated
+// endpoint.
+type paginatedResponse[T any] struct {
+	Page    int `json:"page"    url:"page,omitempty"`
+	Pages   int `json:"pages"   url:"pages,omitempty"`
+	Results int `json:"results" url:"results,omitempty"`
+	Data    []T `json:"data"`
+}
+
+// getPaginatedResults aggregates results from the given
+// paginated endpoint using the provided ListOptions.
+// nolint:funlen
+func getPaginatedResults[T any](
+	ctx context.Context,
+	client *Client,
+	endpoint string,
+	opts *ListOptions,
+) ([]T, error) {
+	var resultType paginatedResponse[T]
+
+	result := make([]T, 0)
+
+	req := client.R(ctx).SetResult(resultType)
+
+	if opts == nil {
+		opts = &ListOptions{PageOptions: &PageOptions{Page: 0}}
+	}
+
+	if opts.PageOptions == nil {
+		opts.PageOptions = &PageOptions{Page: 0}
+	}
+
+	// Apply all user-provided list options to the base request
+	if err := applyListOptionsToRequest(opts, req); err != nil {
+		return nil, err
+	}
+
+	// Makes a request to a particular page and
+	// appends the response to the result
+	handlePage := func(page int) error {
+		req.SetQueryParam("page", strconv.Itoa(page))
+
+		res, err := coupleAPIErrors(req.Get(endpoint))
+		if err != nil {
+			return err
+		}
+
+		response := res.Result().(*paginatedResponse[T])
+
+		opts.Page = page
+		opts.Pages = response.Pages
+		opts.Results = response.Results
```

Comment:
I'm not a big fan of writing to opts here, but this would technically be a breaking change otherwise since users can read from ListOptions afterwards.